File: examples/models1.r
Mb_1 <- function(x, c) {
	#cat("(R) M1: x =", x, " type =", typeof(x), "\n")
	#cat("(R) M1: c =", c, " type =", typeof(c), "\n")
	return(c(c*x, -c*x))
}

Mb_2 <- function(c) {
	return(c+1)
}

Mb_2v2 <- function() {
	return(100)
}

Mb_3 <- function(z) {
	return(c(z*5, -z*5))
}

Mb_4 <- function(z) {
	return(z/5)
}

Mc_1 <- function(x) {
	return(exp(x))
}

Mc_2 <- function(xprime) {
	return(log(xprime))
}

cy_a2b <- function(a) {
	return(a*2);
}

cy_b2a <- function(b) {
	return(b/2);
}

cy_a0 <- function(a0) {
	return(a0 + 1);
}

cy_b0 <- function(b0) {
	return(b0 - 1);
}

cy_d <- function(a, b) {
	return(a + b);
}

File: examples/models1.lua
-- basic models - compute y from either directly from real x and category c
-- or indirectly with intermediate var z

model "Mb_1"
	param "x"
	param "c" check(set(1, 2, 3))

	returns "y"
	impl "R::examples/models1.r::Mb_1"
	returns "y2"
	impl "R::examples/models1.r::Mb_1"

model "Mb_2"
	param "c"

	returns "z"
	impl "R::examples/models1.r::Mb_2"

-- "better" but more specific model for case c=4
model "Mb_2v2"
	-- check c=4 but don't take it as a parameter
	-- the syntax is a bit awkward?
	check(set(4), 0, inf, "c")

	returns "z"
	impl "R::examples/models1.r::Mb_2v2"

model "Mb_3"
	param "z" check(ival(0, inf))

	returns "y"
	impl "R::examples/models1.r::Mb_3"
	returns "y2"
	impl "R::examples/models1.r::Mb_3"

model "Mb_4"
	param "z" check(ival(-inf, 0))

	returns "w"
	impl "R::examples/models1.r::Mb_4"

-- cyclic models - x and x' are "equivalent" in the sense that they can be estimated
-- from each other but one MUST be given

-- TODO (check models1.r): conversion syntax?
-- eg. model M returns exp(x) but model M' takes x as parameter

model "Mc_1"
	param "x"
	returns "x'"

	impl "R::examples/models1.r::Mc_1"

model "Mc_2"
	param "x'"

	returns "x"
	impl "R::examples/models1.r::Mc_2"

-- cyclic models for dijkstra

model "cy_a2b"
	param "a" -- check(ival(-inf, 0))
	returns "b"
	impl "R::examples/models1.r::cy_a2b"

model "cy_b2a"
	param "b"
	returns "a"
	impl "R::examples/models1.r::cy_b2a"

model "cy_a0"
	param "a0"
	returns "a"
	impl "R::examples/models1.r::cy_a0"

model "cy_b0"
	param "b0"
	returns "b"
	impl "R::examples/models1.r::cy_b0"

model "cy_d"
	param "a"
	param "b"
	returns "d"
	impl "R::examples/models1.r::cy_d"
